File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    
    public boolean binarySearch(int[] nums, int target){
        
        int low = 0;
        int high = nums.length - 1;
        
        while(low <= high){
            
            int mid = low + (high - low) / 2;
            
            if(nums[mid] < target){
                low = mid + 1;
            }
            else if(nums[mid] > target){
                high = mid - 1;
            }
            else {
                return true;
            }
        }
        return false;
    }
    
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int rows = matrix.length;
        int cols = matrix[0].length;
        
        int top = 0;
        int bot = rows - 1;
        
        while(top <= bot){
            
            int row = Math.floorDiv((top+bot), 2);
            
            if(matrix[row][matrix[row].length - 1] < target){
                top = row + 1;
            }
            else if(matrix[row][0] > target){
                bot = row - 1;
            }
            else{
                break;
            }
        }
        if(!(top <= bot)){
            return false;
        }
        else{
            return binarySearch(matrix[Math.floorDiv((top+bot), 2)], target);
        }
        
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low = 0
        high = len(matrix) - 1

        rows = 0 
        cols = len(matrix[0]) - 1
        while(low <= high):
            
            mid = (low + high) // 2

            if matrix[mid][0] > target:
                high = mid - 1
            elif matrix[mid][-1] < target:
                low = mid + 1
            else:
                return self.binarySearch(matrix[mid], target)

        return False


    def binarySearch(self, nums: List[int], target: int) -> bool:

        low = 0
        high = len(nums) - 1
        while(low <= high):
            
            mid = (low + high) // 2
            if nums[mid] == target:
                return True
            elif nums[mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return False